Add support for GL_EXT_shared_memory_block

cmarcelo opened the pull request:
This will use generate SPIR-V that uses SPV_KHR_workgroup_memory_explicit_layout extension.

The first patch is a fix for #2400 that I can move to a PR of its own if necessary. It is here because we build on top of that in a later patch.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/KhronosGroup/glslang?pullRequest=2512) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/KhronosGroup/glslang?pullRequest=2512) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/KhronosGroup/glslang?pullRequest=2512) it.</sub>

cmarcelo commented on known_good.json:
```diff
@@ -5,14 +5,14 @@
       "subrepo" : "KhronosGroup/SPIRV-Tools",
       "subdir" : "External/spirv-tools",
-      "commit" : "ad898cb9498c65cac7c5b6f9ebddb610bd0852ad"
+      "commit" : "28b7ca928692ecaaa8558357f9f9ce66e14ac8ef"
```
NOTE: This will be updated once https://github.com/KhronosGroup/SPIRV-Tools/pull/4128 lands.

cmarcelo replied:
Updated.

cmarcelo commented on known_good.json:
```diff
@@ -5,14 +5,14 @@
       "subrepo" : "KhronosGroup/SPIRV-Headers",
       "subdir" : "External/spirv-tools/external/spirv-headers",
-      "commit" : "f027d53ded7e230e008d37c8b47ede7cd308e19d"
+      "commit" : "38fe9b08508c876ba830f250a488c51003531e7e"
```
NOTE: This will be updated once https://github.com/KhronosGroup/SPIRV-Headers/pull/192 lands.

cmarcelo replied:
Updated.

cmarcelo commented:
@greg-lunarg could you take a look at this?

greg-lunarg commented:
I am not seeing 28b7ca as a valid commit for SPIRV-Tools on github. This new value in known_good.json appears to be the reason for the CI failures. You indicate that you are going to update these anyway once SPIRV-Tools and SPIRV-Headers are updated. Once valid github commits are put in here, things should go better.

cmarcelo commented:
> I am not seeing 28b7ca as a valid commit for SPIRV-Tools on github. This new value in known_good.json appears to be the reason for the CI failures. You indicate that you are going to update these anyway once SPIRV-Tools and SPIRV-Headers are updated. Once valid github commits are put in here, things should go better.

@greg-lunarg Correct, these are the hashes of my PRs, but they don't get propagated in a regular clone. SPIRV-Headers already merged and SPIRV-Tools is in the process, I'll update this PR once the latter lands.

I did push this PR earlier because some of the new changes can be reviewed already, e.g. the first patch that fixes #2400. I can also move it to a separate PR of its own if it helps, just let me know.

greg-lunarg commented:
I looked at most of this already but would like to look at the #2400 fix.

alan-baker commented:
I merged the spirv-tools change, but it also includes validation changes for workgroup variable initializers and stricter validation for offset. The test expectations in glslang will need updated as a result.

cmarcelo commented:
Thanks Alan. I'll try to update the test expectations and include that in this PR.

cmarcelo commented:
@alan-baker @greg-lunarg updated.

greg-lunarg commented:
As a whole, these changes look good.

Do we need to squash some of these? I just want to make sure we don't leave points at which the repo is not coherent.

cmarcelo commented:
> As a whole, these changes look good.
> 
> Do we need to squash some of these? I just want to make sure we don't leave points at which the repo is not coherent.

Updated. I've squashed:

- The known_good and the copy of header update, since kind of make sense.
- The three last patches.
